### PR TITLE
feat: add function to manually check rate limit

### LINKED ISSUE #346: Add parser option
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

If you use fastify with trpc you most likely use some kind of parser. We are using fastify + trpc + zod as a parser. There is an option for response `errorResponseBuilder` but it expect an object. So we can't fully specify what we return (in our case it should be either TrpcError or stringified json for zod), so that's the only reason we cant currently use this package. (Unless you can think of an easier solution for this, would be very helpful). It would be nice to have a parser option which would apply specified parser to the response (you can include just a couple, the most popular ones)

### Motivation

_No response_

### Example

_No response_

### PR DESCRIPTION
**Pull request by Charioteer**
Hey fastify contributors,

Following on from issue #346 and my [comment](https://github.com/fastify/fastify-rate-limit/issues/346#issuecomment-2331639190), I finally had some spare time to implement a draft version of my idea as suggested by @mcollina. I realized that any store (child) instance receives all the `options` provided to `fastify.rateLimit` when it is constructed, which is why I had to adjust my initial idea and come up with a slightly different solution.

I did not provide any documentation or additional tests yet (all existing tests run green tho), because I wanted to ask for your feedback first.

Please let me know what you think and as soon as everything looks good to you, I will finalize my PR with proper docs and tests.

Thanks and cheers
Patrick

# How to Use

```ts
import Fastify from "fastify";
import { fastifyRateLimit } from "@fastify/rate-limit";

const fastify = Fastify();

await fastify.register(fastifyRateLimit, {
  global: false,
  max: 10,
  timeWindow: "10 seconds",
});

const checkRateLimit = fastify.createRateLimit(); // this will use the global options provided to fastifyRateLimit

fastify.get("/", async (request, reply) => {
  const limit = await checkRateLimit(request);

  if(!limit.isAllowed && limit.isExceeded) {
    return reply.code(429).send("Limit exceeded");
  }

  return reply.send("Hello world");
});


const checkCustomRateLimit = fastify.createRateLimit({ max: 100 }); // max provided to override global options

fastify.get("/custom", async (request, reply) => {
  const limit = await checkCustomRateLimit(request);

  if(!limit.isAllowed && limit.isExceeded) {
    return reply.code(429).send("Limit exceeded");
  }

  return reply.send("Hello world");
});
```

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by mcollina — COMMENTED** [review]
Looks good! Finish it up (be careful that CI is failing).

**Comment by Charioteer** [comment]
Hi @mcollina,

after @umakantp has closed their PR #410 because of the "authorship problem", I thought of putting more pressure on myself to finally get this done. I added unit tests, type tests, as well as documentation about the usage of this feature. I also merged the latest commit of `fastify-rate-limit` to my fork already to avoid conflicts. LMKWYT.

At this point, I also would like to appreciate @umakantp again for continuing this PR and putting time and effort into it. Please excuse me if I caused any trouble on your side.

**Comment by umakantp** [comment]
Hi @mcollina If you can please take a look and merge if things are good?

**Comment by gurgunday** [comment]
It looks interesting, I'll take a look at this tomorrow

File: index.js [hunk 1]
```diff
@@ -125,16 +125,17 @@ async function fastifyRateLimit (fastify, settings) {
 
   fastify.decorateRequest(pluginComponent.rateLimitRan, false)
 
+  if (!fastify.hasDecorator('createRateLimit')) {
+    fastify.decorate('createRateLimit', (options) => {
+      const args = createLimiterArgs(pluginComponent, globalParams, options)
+      return (req) => applyRateLimit(...args, req)
```
Comment (Eomm) [inline index.js:131]: Instead of decostruct `args` we can:
```suggestion
      return (req) => applyRateLimit.apply(this, args.concat(req))
```

(and even better, avoid the `concat`

**Review by Eomm — COMMENTED** [review]
Technically this change is awesome!

A minor suggestion, but LGTM!

**Review by gurgunday — COMMENTED** [review]
Can you run a benchmark using the basic example in the examples folder with `autocannon`?

https://github.com/fastify/fastify-rate-limit/blob/main/example/example-simple.mjs

**Comment by Charioteer** [comment]
Hi @gurgunday,

sorry for yet anoter late reply...

I wasn't sure which kind of benchmarks you'd expect. I still ran two benchmarks with `autocannon` and its default settings on my machine with my initial version **without** @Eomm's suggestion:
```
Running 10s test @ http://localhost:3000
10 connections


┌─────────┬──────┬──────┬───────┬──────┬─────────┬─────────┬───────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%  │ Avg     │ Stdev   │ Max   │
├─────────┼──────┼──────┼───────┼──────┼─────────┼─────────┼───────┤
│ Latency │ 0 ms │ 0 ms │ 1 ms  │ 1 ms │ 0.05 ms │ 0.29 ms │ 16 ms │
└─────────┴──────┴──────┴───────┴──────┴─────────┴─────────┴───────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬───────────┬──────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg       │ Stdev    │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼───────────┼──────────┼─────────┤
│ Req/Sec   │ 14,959  │ 14,959  │ 18,431  │ 19,039  │ 18,150.91 │ 1,103.65 │ 14,959  │
├───────────┼─────────┼─────────┼─────────┼─────────┼───────────┼──────────┼─────────┤
│ Bytes/Sec │ 4.42 MB │ 4.42 MB │ 6.95 MB │ 7.18 MB │ 6.73 MB   │ 751 kB   │ 4.42 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴───────────┴──────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 11

10000 2xx responses, 189656 non 2xx responses
200k requests in 11.02s, 74 MB read
```

and with @Eomm's suggestion:
```
Running 10s test @ http://localhost:3000
10 connections


┌─────────┬──────┬──────┬───────┬──────┬─────────┬─────────┬───────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%  │ Avg     │ Stdev   │ Max   │
├─────────┼──────┼──────┼───────┼──────┼─────────┼─────────┼───────┤
│ Latency │ 0 ms │ 0 ms │ 1 ms  │ 1 ms │ 0.05 ms │ 0.24 ms │ 18 ms │
└─────────┴──────┴──────┴───────┴──────┴─────────┴─────────┴───────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬───────────┬────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg       │ Stdev  │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼───────────┼────────┼─────────┤
│ Req/Sec   │ 15,791  │ 15,791  │ 18,399  │ 19,247  │ 18,348.73 │ 897.8  │ 15,785  │
├───────────┼─────────┼─────────┼─────────┼─────────┼───────────┼────────┼─────────┤
│ Bytes/Sec │ 4.73 MB │ 4.73 MB │ 6.93 MB │ 7.25 MB │ 6.81 MB   │ 673 kB │ 4.73 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴───────────┴────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 11

10000 2xx responses, 191806 non 2xx responses
202k requests in 11.02s, 74.9 MB read
```

## Machine and Versions
- Machine (MacBook Pro 2019): 
  - CPU: 2,3 GHz 8-Core Intel Core i9
  - GPU: AMD Radeon Pro 5500M 4 GB
  - RAM: 16 GB 2667 MHz DDR4
  - OS: Sonoma 14.5
- Node v22.14.0
- Fastify 5.3.0

**Comment by umakantp** [comment]
Looks like this has fallen behind on the priority list? @gurgunday ?

**Comment by gurgunday** [comment]
Hey, sorry for the late reply, just to get an idea, can you run the same on `master` too? I will approve if there isn't a huge regression

**Comment by Charioteer** [comment]
Hey @gurgunday,

I ran the benchmark again for `main@latest` and for my PR in sequence so that the comparison is as fair as possible. System and versions did not change.

## Master
```
Running 10s test @ http://localhost:3000
10 connections


┌─────────┬──────┬──────┬───────┬──────┬─────────┬─────────┬───────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%  │ Avg     │ Stdev   │ Max   │
├─────────┼──────┼──────┼───────┼──────┼─────────┼─────────┼───────┤
│ Latency │ 0 ms │ 0 ms │ 1 ms  │ 1 ms │ 0.05 ms │ 0.32 ms │ 22 ms │
└─────────┴──────┴──────┴───────┴──────┴─────────┴─────────┴───────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬───────────┬──────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg       │ Stdev    │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼───────────┼──────────┼─────────┤
│ Req/Sec   │ 14,175  │ 14,175  │ 19,119  │ 19,423  │ 18,633.82 │ 1,436.04 │ 14,174  │
├───────────┼─────────┼─────────┼─────────┼─────────┼───────────┼──────────┼─────────┤
│ Bytes/Sec │ 4.12 MB │ 4.12 MB │ 7.21 MB │ 7.32 MB │ 6.91 MB   │ 888 kB   │ 4.12 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴───────────┴──────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 11

10000 2xx responses, 194989 non 2xx responses
205k requests in 11.02s, 76.1 MB read
```

## My PR
```
Running 10s test @ http://localhost:3000
10 connections


┌─────────┬──────┬──────┬───────┬──────┬─────────┬─────────┬───────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%  │ Avg     │ Stdev   │ Max   │
├─────────┼──────┼──────┼───────┼──────┼─────────┼─────────┼───────┤
│ Latency │ 0 ms │ 0 ms │ 1 ms  │ 1 ms │ 0.05 ms │ 0.32 ms │ 23 ms │
└─────────┴──────┴──────┴───────┴──────┴─────────┴─────────┴───────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬───────────┬──────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg       │ Stdev    │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼───────────┼──────────┼─────────┤
│ Req/Sec   │ 15,399  │ 15,399  │ 18,159  │ 19,151  │ 18,119.64 │ 1,019.03 │ 15,392  │
├───────────┼─────────┼─────────┼─────────┼─────────┼───────────┼──────────┼─────────┤
│ Bytes/Sec │ 4.58 MB │ 4.58 MB │ 6.85 MB │ 7.22 MB │ 6.72 MB   │ 707 kB   │ 4.58 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴───────────┴──────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 11

10000 2xx responses, 189305 non 2xx responses
199k requests in 11.02s, 73.9 MB read
```

## TL;DR
My PR has about <ins>**2.9%** less throughput</ins>. Let me know if that is too much regression. I did not check whether there is room for improvement yet.

**Review by gurgunday — APPROVED** [review]
I think 2% is acceptable, we should be able to improve this further with some optimizations but it lgtm currently for this PR

**Review by Fdawgs — CHANGES_REQUESTED** [review]
Blocking until I can find time to review, from brief glance the docs could be refined.

File: README.md [hunk 2]
```diff
@@ -450,6 +450,80 @@ fastify.get('/private', {
...
+### Manual Rate Limit
+
+In some situations, the default behavior of this plugin may not be sufficient and you want to implement your own rate limit strategy.
+This might be the case as well if you want to integrate it with other technologies such as [GraphQL](https://graphql.org/) or [tRPC](https://trpc.io/).
+You can create your own limiter function with `fastify.createRateLimit()`.
+By default, this limiter function uses the global [options](#options) you defined during plugin registration.
+You can override some or all of your global options by passing them to `createRateLimit()`.
+The following options can be overridden: `store`, `skipOnError`, `max`, `timeWindow`, `allowList`, `keyGenerator`, and `ban`.
+
+The next example demonstrates the usage of a custom rate limiter:
```
Comment (Fdawgs) [inline README.md:462]:
```suggestion
A custom limiter function can be created with `fastify.createRateLimit()`, which is handy when needing to integrate with
technologies like [GraphQL](https://graphql.org/) or [tRPC](https://trpc.io/). This function uses the global [options](#options) set
during plugin registration, but you can override options such as `store`, `skipOnError`, `max`, `timeWindow`,
`allowList`, `keyGenerator`, and `ban`.

Example usage:
```

File: README.md [hunk 3]
```diff
@@ -450,6 +450,80 @@ fastify.get('/private', {
...
+    return reply.code(429).send("Limit exceeded");
+  }
+
+  return reply.send("Hello world");
+});
+```
+
+A custom limiter function created with `fastify.createRateLimit()` only requires a `FastifyRequest` as the first parameter:
+
+```ts
```
Comment (Fdawgs) [inline README.md:507]:
```suggestion
```js
```

**Review by Fdawgs — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -450,6 +450,78 @@ fastify.get('/private', {
...
+Example usage:
+
+```js
+import Fastify from 'fastify'
+
+const fastify = Fastify()
+
+// register with global options
+await fastify.register(import('@fastify/rate-limit'), {
+  max: 100,
```
Comment (Eomm) [inline README.md:469]: without it, routes will have 2 rate limit hooks

```suggestion
  global : false,
  max: 100,
```

File: README.md [hunk 5]
```diff
@@ -450,6 +450,78 @@ fastify.get('/private', {
...
+  timeWindow: '1 minute'
+})
+
+ // checkRateLimit will use the global options provided above when called
```
Comment (Eomm) [inline README.md:473]:
```suggestion
// checkRateLimit will use the global options provided above when called
```

File: index.js [hunk 6]
```diff
@@ -125,16 +125,17 @@ async function fastifyRateLimit (fastify, settings) {
...
-      if (typeof options === 'object') {
-        const newPluginComponent = Object.create(pluginComponent)
-        const mergedRateLimitParams = mergeParams(globalParams, options, { routeInfo: {} })
-        newPluginComponent.store = newPluginComponent.store.child(mergedRateLimitParams)
-        return rateLimitRequestHandler(newPluginComponent, mergedRateLimitParams)
-      }
-
-      return rateLimitRequestHandler(pluginComponent, globalParams)
+      const args = createLimiterArgs(pluginComponent, globalParams, options)
+      return rateLimitRequestHandler(...args)
```
Comment (Eomm) [inline index.js:138]:
```suggestion
      return rateLimitRequestHandler.apply(this, args)
```

Comment (Fdawgs) [reply index.js:138]: Enlighten me @Eomm, is apply faster than the spread?

Comment (Eomm) [reply index.js:138]: It was not for perf, but for the `this` context - but yeah, rechecked it and we don't need it!

**Review by Eomm — APPROVED** [review]
